### PR TITLE
Add the estimates tibble

### PR DESCRIPTION
**Pull request by Bai-Li-NOAA**
# What is the feature?
* Add the estimates tibble by merging TMB and FIMS JSON outputs

# How have you implemented the solution?
* feat(estimates tibble):
  * Add reshape_tmb_estimates() and reshape_json_estimates() to convert
    TMB and FIMS JSON outputs into tibbles.
  * Add reshape_json_derived_quantities() to convert derived quantities
    from JSON output into a tibble.
  * Merge the two tibbles in FIMSFit() to maximize extracted information.
* fix(finalize):
  * Remove the trailing comma in rcpp_fleet.hpp JSON output. Conditionally
    add a comma if nlengths > 0 to ensure valid JSON syntax, preventing
    errors when log_q is the last element. @msupernaw, could you review the changes?
* fix(Rcpp):
  * Use std::stringstream to construct parameter names with additional
    context in TMB output (module name, parameter name, module ID, and
    unique parameter ID).
  * Remove unused estimated_log_M and estimated_log_init_naa from
    inst/include/interface/rcpp/rcpp_objects/rcpp_population.hpp. @msupernaw, 
    could you review the changes?
* refactor(tests):
  * Add a deterministic run to setup.R and ignore its results in .gitignore.
  * Update tests to reflect the latest estimate tibble naming conventions.
  * Add tests for new functionality, including get_*() methods in FIMSFit().

# Does the PR impact any other area of the project, maybe another repo?
* Other branches may need to update certain tests to use the estimate tibble. This could conflict with ongoing JSON output work in those branches.

Note: 
* This PR addresses the suggestions made before April 14, 2025, in discussion #771. These include: 1) removing the parameter_min and parameter_max columns, 2) renaming log_lik to log_like, and log_lik_cv to log_like_cv.
Comments related to random effects will be left as-is for now and can be revisited as part of @Andrea-Havron-NOAA's ongoing work on random effects.
* The derived quantities section in the estimates tibble is still a work in progress, as the TMB and JSON outputs currently provide different information.

### CODE REVIEW

**Comment by Bai-Li-NOAA** [comment]
@MOshima-PIFSC, the PR is ready for review. Feel free to let me know if you won’t have time to look at it this week.

File: R/fimsfit.R [hunk 1]
```diff
@@ -385,27 +385,137 @@ FIMSFit <- function(
...
+    # Reshape the TMB estimates
+    tmb_estimates <- reshape_tmb_estimates(
+      obj = obj, 
+      sdreport = sdreport,
+      opt = opt,
+      parameter_names = parameter_names
+    )
+ 
+    # Create JSON output for FIMS run
+    finalized_fims <- finalize(opt$par, obj$fn, obj$gr)
```
Comment (kellijohnson-NOAA) [inline R/fimsfit.R:403]:
```suggestion
    finalized_fims <- finalize(opt[["par"]], obj[["fn"]], obj[["gr"]])
```

File: R/create_default_parameters.R [hunk 2]
```diff
@@ -13,7 +13,7 @@
 #'   applied for each fleet that is not specified but present in `data` based
 #'   on the types of information present for that fleet.
 #' @param recruitment A list specifying the settings for the recruitment
-#'   module. The default is a Beverton--Holt recruitment relationship with
+#'   module. The default is a Beverton-Holt recruitment relationship with
```
Comment (kellijohnson-NOAA) [inline R/create_default_parameters.R:16]: @Bai-Li-NOAA I am curious why you made this and other associated changes. An en-dash is appropriate between last names of individuals and between contrasting comparisons/relationships. I cannot remember if roxygen parses these into true en-dashes, i.e., a dash the width of an "n" rather than two dashes that are used to approximate the width of an n and differentiate it from a regular dash or an em-dash.

File: R/fimsfit.R
Comment (MOshima-PIFSC) [inline R/fimsfit.R:1]: nit pick: small typo in line 322 (I know this isn't part of your changes), "A object" should be "An object"

**Review by MOshima-PIFSC — COMMENTED** [review]
I've reviewed the files and overall it looks really good @Bai-Li-NOAA! I had a few minor nit picks and suggestions for you to look at but other than that, I have no issues with the changes.

File: R/fimsfit.R [hunk 3]
```diff
@@ -418,6 +528,8 @@ FIMSFit <- function(
     number_of_parameters = number_of_parameters,
     timing = timing,
     version = version
+    # json_derived_quantities = json_derived_quantities,
+    # json_density_component = json_density_component
```
Comment (kellijohnson-NOAA) [inline R/fimsfit.R:532]: I am assuming these lines can be removed?

File: deterministic_run.csv
Comment (kellijohnson-NOAA) [inline deterministic_run.csv:1]: Are these three csv files supposed to be included?

File: R/create_default_parameters.R (hunk 2 above)
Comment (Bai-Li-NOAA) [reply R/create_default_parameters.R:16]: @kellijohnson-NOAA. Thanks for explaining the reason behind the use of `--`. I made the changes initially because when I printed the `estimates` tibble in R, it displayed `Beverton--Holt`, which I assumed was a typo. I'll revisit the changes I made and also review the rendered files to confirm how it’s being parsed.

File: R/fimsfit.R
Comment (Bai-Li-NOAA) [reply R/fimsfit.R:1]: Good catch, thanks!

File: R/reshape_output.R [hunk 4]
```diff
@@ -0,0 +1,257 @@
...
+            module_type = ..4
+          )
+      }
+    }
+  ) |>
+    # Combine all the processed tibbles into a single tibble by stacking rows.
+    dplyr::bind_rows() |>
+    # Reorder the columns to place `module_name`, `module_id`, and `module_type` at the beginning.
+    dplyr::relocate(module_name, module_id, module_type, .before = everything()) |>
+    # Remove columns `parameter_min` and `paramter_max` as we currntly don't
```
[the remark this answers is not in the document]
Comment (Bai-Li-NOAA) [reply R/reshape_output.R:94]: Thanks! Fixed the typo.

File: R/reshape_output.R [hunk 5]
```diff
@@ -0,0 +1,257 @@
...
+    label = character(),
+    # The unique ID of the parameter
+    parameter_id = integer(),
+    # The fleet name associated with the parameter or derived quantity
+    fleet_name = character(),
+    # The age associated with the parameter or derived quantity
+    age = numeric(),
+    # The length associated with the parameter or derived quantity
+    length = numeric(),
+    # The modeled time perioed that the value pertains to
```
[the remark this answers is not in the document]
Comment (Bai-Li-NOAA) [reply R/reshape_output.R:133]: Thanks! Fixed the typo.

File: R/reshape_output.R [hunk 6]
```diff
@@ -0,0 +1,257 @@
...
+    time = integer(),
+    # The initial value use to start the optimization procedure
+    initial = numeric(),
+    # The estaimted parameter value, which would be the MLE estimate or the value
```
[the remark this answers is not in the document]
Comment (Bai-Li-NOAA) [reply R/reshape_output.R:137]: Thanks! Fixed the typo.

File: tests/testthat/test-get_estimates.R [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+      # Remove the estimate, uncertainty, and gradient columns, as they
+      # may change between runs
+      dplyr::select(-estimate, -uncertainty, -gradient) |>
+      print(n = 320, width = Inf)
+  )
+})
+
+## Edge handling ----
+test_that("get_estimates() returns correct outputs for edge cases", {
+  #' @description Test that get_estimates("invalide_fit") returns an error.
```
[the remark this answers is not in the document]
Comment (Bai-Li-NOAA) [reply tests/testthat/test-get_estimates.R:82]: Done. Thanks!

File: R/reshape_output.R [hunk 8]
```diff
@@ -0,0 +1,257 @@
...
+    parameter_id = integer(),
+    # The fleet name associated with the parameter or derived quantity
+    fleet_name = character(),
+    # The age associated with the parameter or derived quantity
+    age = numeric(),
+    # The length associated with the parameter or derived quantity
+    length = numeric(),
+    # The modeled time perioed that the value pertains to
+    time = integer(),
+    # The initial value use to start the optimization procedure
```
Comment (kellijohnson-NOAA) [inline R/reshape_output.R:135]: These four columns are NA in the output, which is fine for now but is there a plan in how to fill them in the future?

File: R/reshape_output.R [hunk 9]
```diff
@@ -0,0 +1,257 @@
...
+  estimates <- estimates |>
+    # Split labels and extract module, id, label, and parameter id
+    dplyr::mutate(label_splits = strsplit(label, split = "\\.")) |>
+    dplyr::rowwise() |>
+    dplyr::mutate(
+      module_name = ifelse(length(label_splits) > 1, label_splits[[1]], NA_character_),
+      module_id = ifelse(length(label_splits) > 1, as.integer(label_splits[[3]]), NA_integer_),
+      label = ifelse(length(label_splits) > 1, label_splits[[2]], label),
+      parameter_id = ifelse(length(label_splits) > 1, as.integer(label_splits[[4]]), NA_integer_)
+    ) |>
```
Comment (kellijohnson-NOAA) [inline R/reshape_output.R:199]: Just a note that this section of code might be easier with `tidyr::separate_wider_*()` functions.

**Review by kellijohnson-NOAA — APPROVED** [review]
Thanks @Bai-Li-NOAA for all of this, the estimates tibble is off to a great start. I approve these changes, when you are done editing the PR please let me know. I will then, squash all the commits, rebase to dev, merge them into dev, and then rebase Nathan's PR.

File: R/create_default_parameters.R (hunk 2 above)
Comment (Bai-Li-NOAA) [reply R/create_default_parameters.R]: I've reverted the change from Beverton--Holt to Beverton-Holt. Thanks for pointing that out.

File: R/fimsfit.R [hunk 10]
```diff
@@ -385,27 +385,137 @@ FIMSFit <- function(
...
+      obj = obj, 
+      sdreport = sdreport,
+      opt = opt,
+      parameter_names = parameter_names
     )
+    # Create JSON output for FIMS run
+    finalized_fims <- finalize(obj$par, obj$fn, obj$gr)
+    # Reshape the JSON estimates 
+    json_estimates <- reshape_json_estimates(finalized_fims, opt)
   }
```
[the remark this answers is not in the document]
Comment (Bai-Li-NOAA) [reply R/fimsfit.R]: Great suggestions! I've removed the duplicated code and added more comments.

File: R/fimsfit.R [hunk 11]
```diff
@@ -385,27 +385,137 @@ FIMSFit <- function(
...
+  # Merge json_derived_quantities into estimates based on common columns
+  estimates <- dplyr::full_join(
+    estimates,
+    json_derived_quantities,
+    by = dplyr::join_by(
+      derived_quantity_id,
+      label
+    )
+  ) |>
+    dplyr::mutate(
```
[the remark this answers is not in the document]
Comment (Bai-Li-NOAA) [reply R/fimsfit.R:495]: Great suggestion! I've added additional comments to the codebase.

File: R/fimsfit.R (hunk 3 above)
Comment (Bai-Li-NOAA) [reply R/fimsfit.R]: Yes. These lines have been removed!

File: deterministic_run.csv
Comment (Bai-Li-NOAA) [reply deterministic_run.csv:1]: I've removed the CSV files from the codebase.

File: R/reshape_output.R (hunk 8 above)
Comment (Bai-Li-NOAA) [reply R/reshape_output.R]: Yes. Matthew has started adding information to the JSON output in the dev-model-families branch. We can populate these columns once the dev-model-families branch is merged. I've added a TODO comment as a reminder.
